### DISCUSSION TITLE
How to convert string to ReadOnlyMemory<byte>?

### DISCUSSION THREAD

**Discussion by rabbitLove520** [rabbitmq/rabbitmq-dotnet-client#1273] [General]
```
public static void BasicPublish(this IModel model, string exchange, string routingKey, IBasicProperties basicProperties, ReadOnlyMemory<byte> body);
```
via `str.AsMemory()`, I got `ReadOnlyMemory<char>`, any method to get `ReadOnlyMemory<byte>`?

**Comment by lukebakken** [comment]
https://github.com/rabbitmq/rabbitmq-dotnet-client/search?q=BasicPublish&type=code

https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/99a2b6948538f7ac0cb975ec97f21a5ac59edce8/projects/Unit/TestBasicPublish.cs#L23-L35

**Reply by rabbitLove520** [reply]
`sendBody` is `byte[]`, not `ReadOnlyMemory<byte>`

**Reply by stebet** [reply]
> `sendBody` is `byte[]`, not `ReadOnlyMemory<byte>`

A `byte[]` is implicitly convertible to `ReadOnlyMemory<byte>`.

See: https://github.com/dotnet/runtime/blob/060a1967649469fa71986dc92bb41629c0970946/src/libraries/System.Private.CoreLib/src/System/ReadOnlyMemory.cs#L114

**Reply by lukebakken** [reply]
@pigLoveRabbit520 I pointed you to code that we use to test this library. Why would you think that it doesn't work?

Please, when someone offers you assistance, **try** their suggestion first before questioning it.

**Reply by rabbitLove520** [reply]
> > `sendBody` is `byte[]`, not `ReadOnlyMemory<byte>`
> 
> A `byte[]` is implicitly convertible to `ReadOnlyMemory<byte>`.
> 
> See: https://github.com/dotnet/runtime/blob/060a1967649469fa71986dc92bb41629c0970946/src/libraries/System.Private.CoreLib/src/System/ReadOnlyMemory.cs#L114
thks, I noticed it. `Encoding.UTF8.GetBytes()` is a copy of string, I want to avoid copying to get byte[] of string.

**Reply by stebet** [reply]
You'll have to make a copy of it since there is no easy way to get a byte reference to an internal .NET string reference. .NET stores all string as UTF16 and can shuffle them around.

It IS possible to do through some very nasty unmanaged code and interop and I'd strongly recommend you avoid that at all cost.
